# January Puzzle Solution

I approached this puzzle by using some number-theory and followed by the some simple written code. 
My basic idea was:

- use the special number `12,345,679` to generate possible candites for the puzzle,
- and then identify the unclued middle row as `283,950,617`.


## Why `12,345,679` appears

Let `S` be the sum of the unique digits (note that we can only use 9 same digits everywhere), and let `d` be the GCD of the 9-digit numbers formed by the 9 rows.

Because the rows are cyclic shifts of one another, their sum is

`111,111,111 * S = 3 * 3 * 37 * 333,667 * S`.

So any common divisor has to divide that number.  One of them is in fact a combination of 8 unique digits:

`37 * 333,667 = 12,345,679`.

This number is useful because multiplying it by `1` through `8` produces digit patterns with strong cyclic behavior:

- `12,345,679 * 1 = 12,345,679`
- `12,345,679 * 2 = 24,691,358`
- `12,345,679 * 3 = 37,037,037`
- `12,345,679 * 4 = 49,382,716`
- `12,345,679 * 5 = 61,728,395`
- `12,345,679 * 6 = 74,074,074`
- `12,345,679 * 7 = 86,419,753`
- `12,345,679 * 8 = 98,765,432`

Among these, the distinct-digit values were the ones worth checking against the Sudoku clues.


In [ ]:
BASE = 12_345_679


def has_distinct_digits(n: int) -> bool:
    s = str(n)
    return len(set(s)) == len(s)


multiples = [(k, BASE * k) for k in range(1, 9)]
distinct_candidates = [(k, value) for k, value in multiples if has_distinct_digits(value)]

distinct_candidates


[(1, 12345679),
 (2, 24691358),
 (4, 49382716),
 (5, 61728395),
 (7, 86419753),
 (8, 98765432)]

The distinct-digit multiples are:

- `12,345,679`
- `24,691,358`
- `49,382,716`
- `61,728,395`
- `86,419,753`
- `98,765,432`

At this point I only have to check a tiny number of structured candidates, and `61,728,395` is the one that leads to the Sudoku row cycle I needed.




## Building the Sudoku rows

From there, I used the cyclic family generated from

`395,061,728`.

Its cyclic shifts are exactly the row values in the solved grid.


In [ ]:
def cyclic_shifts(s: str):
    return [s[i:] + s[:i] for i in range(len(s))]


seed = '395061728'
rows = cyclic_shifts(seed)
rows


In [ ]:
for index, row in enumerate(rows, start=1):
    print(f'Row {index}: {row}')


These rows are:

- `395061728`
- `950617283`
- `506172839`
- `061728395`
- `617283950`
- `172839506`
- `728395061`
- `283950617`
- `839506172`

I then reordered them so that they satisfied the given clues. In that arrangement, the unclued middle row is

`283,950,617`.
